src/events.rs

```diff
@@ -0,0 +1,319 @@
+//! Everything related to watching files for a creations, modifications,
+//! deletions, etc.
+
+use std::collections::{HashMap, HashSet};
+use std::fmt::{self, Debug, Formatter};
+use std::io;
+use std::path::{Path, PathBuf};
+use std::pin::Pin;
+use std::task;
+
+use futures_util::pin_mut;
+use futures_util::stream::{Stream as FuturesStream, StreamExt};
+use notify;
+use thiserror::Error;
+use tokio::sync::mpsc;
+
+/// Manages filesystem event watches, and can be polled to receive new events.
+///
+/// Internally, `MuxedEvents` contains a [`notify::Watcher`] from where
+/// filesystem events are proxied. Functionality such as async/await support,
+/// and nonexistent file registration are added.
+///
+/// [`notify::Watcher`]: ../notify/trait.Watcher.html
+pub struct MuxedEvents {
+    inner: notify::RecommendedWatcher,
+    watched_directories: HashMap<PathBuf, usize>,
+    /// Files that are successfully being watched
+    watched_files: HashSet<PathBuf>,
+    /// Files that don't exist yet, but will start once a create event comes
+    /// in for the watched parent directory.
+    pending_watched_files: HashSet<PathBuf>,
+    event_stream: mpsc::UnboundedReceiver<Result<notify::Event, notify::Error>>,
+}
+
+impl Debug for MuxedEvents {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("MuxedEvents")
+            .field("watched_directories", &self.watched_directories)
+            .field("watched_files", &self.watched_files)
+            .field("pending_watched_files", &self.pending_watched_files)
+            .finish()
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Failed to add path to watch")]
+    AddFailure,
+    #[error("Failed to remove path from watch")]
+    RemoveFailure,
+    #[error("Watch already exists for path")]
+    WatchExists,
+    #[error("Error receiving event: {0}")]
+    Event(#[from] std::io::Error),
+}
+
+impl MuxedEvents {
+    /// Constructs a new `MuxedEvents` instance.
+    pub fn new() -> Self {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let inner: notify::RecommendedWatcher = notify::Watcher::new_immediate(move |res| {
+            // The only way `send` can fail is if the receiver is dropped,
+            // and `MuxedEvents` controls both. `unwrap` is not used,
+            // however, since `Drop` idiosyncrasies could otherwise result
+            // in a panic.
+            let _ = tx.send(res);
+        })
+        .expect("Failed to start watcher");
```

Comment:
What are the failure cases for `Watcher`? I'm wondering if this should be returned as an error from `new` instead.

Reply:
Ah, yeah I definitely want to propagate this. Most/all are `std::io` anyway, so it should be relatively easy to normalize.